text/0000-placement-by-return.md

```diff
@@ -0,0 +1,1097 @@
+            return []; // returning constant
+        }
+    }
+    let n = 100;
+    [1; n] // returning variable-length-array expression
+}
+
+// desugar with_multiple_exit_points
+enum __WithMultipleExitPoints__Internal {
+  S0(),
+  S1(i32, usize),
+}
+impl __WithMultipleExitPoints__Internal {
+  unsafe fn start(state: *mut __WithMultipleExitPoints__Internal) -> Layout {
+    while keep_going() {
+        if got_cancelation() {
+            *state = __WithMultipleExitPoints__Internal::S0();
+            return Layout::for_value(&[]);
+        }
+    }
+    let n = 100;
+    *state = __WithMultipleExitPoints__Internal::S1(1, n);
+    Layout::from_size_align_unchecked(n * mem::size_of::<i32>(), mem::align_of::<i32>())
+  }
+  unsafe fn finish(state: *mut __WithMultipleExitPoints__Internal, slot: *mut u8) -> &[i32] {
+    match *state {
+        __WithMultipleExitPoints__Internal::S0() => {
+            slice::from_raw_parts_mut(slot, 0)
+        }
+        __WithMultipleExitPoints_Internal::S1(value, n) => {
+            for i in 0..n { ptr::write(slot.offset(i), value) };
+            slice::from_raw_parts_mut(slot, n)
+        }
+    }
+  }
+}
+```
+
+## How this works with sized types
+
+Any function that returns a sized type can be trivially adapted to the "unsized return generator" interface of `start()` and `finish()`, by having start return a constant and have finish do all the work.
+
+```rust
+fn return_a_sized_value() -> i32 {
+    println!("got random value");
+    4 // determined by a fair dice roll
+}
+
+// desugar return_a_sized_value
+// this is written in order to aid understanding, not because these are good APIs
+use std::alloc::Layout;
+struct __ReturnASizedValue__Internal;
+impl __ReturnASizedValue__Internal {
+  unsafe fn start(state: *mut __WithBranch__Internal) -> Layout {
+    // there's only one possible layout for sized types; that's the definition of being "sized"
+    Layout::new::<i32>()
+  }
+  unsafe fn finish(state: *mut __MyFunction__Internal, slot: *mut u8) -> &mut i32 {
+    // just invoke the function and copy its return value into the slot
+    // this mostly gets the behavior we want, but see below for info about copy elision
+    ptr::copy(&return_a_sized_value(), slot, mem::size_of::<i32>());
+    slot as *mut i32 as &mut i32
+  }
+}
+```
+
+This is how it would conceptually work whenever a facility designed to handle placement is invoked on a function that returns a sized value. This is not how it should be implemented, though, because we don't want a bunch of unnecessary copying.
+
+## Absolutely minimum viable copy elision
+
+To make sure this functionality can be used with no overhead, the language should guarantee some amount of copy elision. The following operations should be guaranteed zero-copy:
+
+* Directly returning the result of another function that also returns the same type.
+* Blocks, unsafe blocks, and branches that have acceptable expressions in their tail position.
```

Comment:
Nit, this should be the last item in the list because it refers to the whole list.